src/js/live-tracker.js

```diff
@@ -0,0 +1,204 @@
+
+    // we can only tell if we are behing live, when seekable changes
+    // once we detect that seekable has changed we check the new seek
+    // end against current time, with a fudge value of half a second.
+    if (newSeekEnd !== this.lastSeekEnd_) {
+      if (this.lastSeekEnd_) {
+        this.segmentLength_ = newSeekEnd - this.lastSeekEnd_;
+      }
+
+      this.pastSeekEnd_ = 0;
+      this.lastSeekEnd_ = newSeekEnd;
+      this.trigger('seekableendchange');
+    }
+
+    this.pastSeekEnd_ = this.pastSeekEnd() + 0.03;
+
+    if (this.isBehind() !== this.behindLiveEdge()) {
+      this.behindLiveEdge_ = this.isBehind();
+      this.trigger('liveedgechange');
+    }
+  }
+
+  /**
+   * handle a durationchange event on the player
+   * and start/stop tracking accordingly.
+   */
+  handleDurationchange() {
+    if (this.player_.duration() === Infinity) {
+      this.start();
+    } else {
+      this.stop();
+    }
+  }
+
+  /**
+   * start tracking live playback
+   */
+  start() {
+    if (this.started()) {
+      return;
+    }
+
+    this.trackingInterval_ = this.setInterval(this.trackLive_, 30);
+    this.trackLive_();
+  }
+
+  /**
+   * stop tracking live playback
+   */
+  stop() {
+    if (!this.started()) {
+      return;
+    }
+    this.pastSeekEnd_ = 0;
+    this.lastSeekEnd_ = null;
+    this.behindLiveEdge_ = null;
+
+    this.clearInterval(this.trackingInterval_);
+    this.trackingInterval_ = null;
+  }
+
+  /**
+   * A helper to get the player seekable end
+   * so that we don't have to null check everywhere
+   */
+  seekableEnd() {
+    const seekable = this.player_.seekable();
+
+    if (!seekable || !seekable.length) {
+      return Infinity;
+    }
+
+    return seekable.end(0);
+  }
+
+  /**
+   * A helper to get the player seekable start
+   * so that we don't have to null check everywhere
+   */
+  seekableStart() {
+    const seekable = this.player_.seekable();
+
+    if (!seekable || !seekable.length) {
+      return 0;
+    }
+
+    return seekable.start(0);
+  }
+
+  /**
+   * Get the live time window
+   */
+  liveWindow() {
+    return this.seekableEnd() - this.seekableStart();
+  }
+
+  /**
+   * Determines if the player is live, only checks wether this component
+   * is tracking live playback or not
+   */
+  isLive() {
+    return this.started();
+  }
+
+  /**
+   * Determines if currentTime is at the live edge and won't fall behind
+   * on each seekableendchange
+   */
+  atLiveEdge() {
+    return !this.behindLiveEdge();
+  }
+
+  /**
+   * get what we expect the live current time to be
+   */
+  liveCurrentTime() {
+    return this.pastSeekEnd() + this.seekableEnd();
+  }
+
+  /**
+   * Returns how far past seek end we expect current time to be
+   */
+  pastSeekEnd() {
+    return this.pastSeekEnd_ || 0;
+  }
+
+  /**
+   * If we are currently behind the live edge, aka currentTime will be
+   * behind on a seekableendchange
+   */
+  behindLiveEdge() {
+    return this.behindLiveEdge_;
```

Comment:
why have both `isBehind()` and this?

Reply:
So that we can fire a `liveedgechange` event, and so that we return the current state rather than a possible state. We could make isBehind_ private.

Reply:
seems to make sense for one of two to not be exposed